### PR TITLE
Make spdx SBOMs readable by trivy scanner

### PR DESCRIPTION
**Pull request by josegomezr**
Before:
```
trivy sbom --format template --template '{{ $vuln := 0 }}{{- range . }}{{- range .Vulnerabilities }}{{- $vuln = add $vuln 1 }}{{- end -}}{{- end -}}Vuln count: {{ $vuln }}'  /home/scc/code/obs-build/obs-sbom.spdx.json
2024-11-18T17:50:14+01:00       INFO    [vuln] Vulnerability scanning is enabled
2024-11-18T17:50:15+01:00       INFO    Detected SBOM format    format="spdx-json"
2024-11-18T17:50:18+01:00       WARN    [sbom] Ignore the OS package as no OS is detected.      file_path="/home/scc/code/obs-build/obs-sbom.spdx.json"
2024-11-18T17:50:18+01:00       INFO    Number of language-specific files       num=1
2024-11-18T17:50:18+01:00       INFO    [gobinary] Detecting vulnerabilities...
2024-11-18T17:50:18+01:00       WARN    Version matching error  err="version error (go1.17.9): malformed version: go1.17.9"
# the same, but 103 times more
2024-11-18T17:50:18+01:00       WARN    Version matching error  err="version error (go1.17.9): malformed version: go1.17.9"
Vuln count: 0
```

After:

```
trivy sbom --format template --template '{{ $vuln := 0 }}{{- range . }}{{- range .Vulnerabilities }}{{- $vuln = add $vuln 1 }}{{- end -}}{{- end -}}Vuln count: {{ $vuln }}'  /home/scc/code/obs-build/obs-sbom.spdx.json
2024-11-18T17:57:43+01:00       INFO    [vuln] Vulnerability scanning is enabled
2024-11-18T17:57:44+01:00       INFO    Detected SBOM format    format="spdx-json"
2024-11-18T17:57:47+01:00       INFO    Detected OS     family="sles" version="15.4"
2024-11-18T17:57:47+01:00       WARN    Unsupported os  family="sles"
2024-11-18T17:57:47+01:00       INFO    Number of language-specific files       num=1
2024-11-18T17:57:47+01:00       INFO    [gobinary] Detecting vulnerabilities...
2024-11-18T17:57:47+01:00       WARN    Using severities from other vendors for some vulnerabilities. Read https://aquasecurity.github.io/trivy/dev/docs/scanner/vulnerability#severity-selection for details.
Vuln count: 56
```

### CODE REVIEW

File: Build/IntrospectGolang.pm [hunk 1]
```diff
@@ -168,6 +168,8 @@ sub buildinfo  {
   my ($fh) = @_;
   my ($vers, $mod) = rawbuildinfo($fh);
   return undef unless defined $vers;
+  $vers =~ s/^go(.*)/\1/;
+
```
Comment (mlschroe) [inline Build/IntrospectGolang.pm:172]: Please use $1 instead of \1

File: generate_sbom [hunk 2]
```diff
@@ -935,6 +940,17 @@ sub spdx_encode_header {
...
+sub spdx_encode_dist {
+  my ($dist) = @_;
+
+  return spdx_encode_pkg({
+    NAME => $dist->{id},
+    VERSION => $dist->{version_id},
+    spdx_id => sprintf('SPDXRef-OperatingSystem-%s', gen_pkg_id($dist)),
+    external_refs => 0
+  }, undef, undef, {});
+
```
Comment (mlschroe) [inline generate_sbom:952]: You need to have OPERATING-SYSTEM in there

File: Build/IntrospectGolang.pm (hunk 1 above)
Comment (josegomezr) [reply Build/IntrospectGolang.pm:172]: Done

File: generate_sbom (hunk 2 above)
Comment (josegomezr) [reply generate_sbom:955]: Added!

**Comment by josegomezr** [comment]
After testing a bleeding edge trivy build after aquasecurity/trivy#7965 and the code on this PR, trivy scanner can finally detect the OS on the generated SPDX and CycloneDX reports properly, including the lifecycle information!

```
$ ../trivy/main sbom  --format template --template '{{ $vuln := 0 }}{{- range . }}{{- range .Vulnerabilities }}{{- $vuln = add $vuln 1 }}{{- end -}}{{- end -}}Vuln count: {{ $vuln }}' obs-sbom.cdx.json 
2024-11-22T22:10:06+01:00       INFO    [vuln] Vulnerability scanning is enabled
2024-11-22T22:10:06+01:00       INFO    Detected SBOM format    format="cyclonedx-json"
2024-11-22T22:10:06+01:00       WARN    Third-party SBOM may lead to inaccurate vulnerability detection
2024-11-22T22:10:06+01:00       WARN    Recommend using Trivy to generate SBOMs
2024-11-22T22:10:06+01:00       INFO    Detected OS     family="sles" version="15.4"
2024-11-22T22:10:06+01:00       INFO    [sles] Detecting vulnerabilities...     os_version="15.4" pkg_num=370
2024-11-22T22:10:06+01:00       INFO    Number of language-specific files       num=1
2024-11-22T22:10:06+01:00       INFO    [gobinary] Detecting vulnerabilities...
2024-11-22T22:10:06+01:00       WARN    Using severities from other vendors for some vulnerabilities. Read https://aquasecurity.github.io/trivy/dev/docs/scanner/vulnerability#severity-selection for details.
2024-11-22T22:10:06+01:00       WARN    This OS version is no longer supported by the distribution      family="sles" version="15.4"
2024-11-22T22:10:06+01:00       WARN    The vulnerability detection may be insufficient because security updates are not provided
Vuln count: 487

$ ../trivy/main sbom  --format template --template '{{ $vuln := 0 }}{{- range . }}{{- range .Vulnerabilities }}{{- $vuln = add $vuln 1 }}{{- end -}}{{- end -}}Vuln count: {{ $vuln }}' obs-sbom.spdx.json 
2024-11-22T22:10:12+01:00       INFO    [vuln] Vulnerability scanning is enabled
2024-11-22T22:10:13+01:00       INFO    Detected SBOM format    format="spdx-json"
2024-11-22T22:10:16+01:00       INFO    Detected OS     family="sles" version="15.4"
2024-11-22T22:10:16+01:00       INFO    [sles] Detecting vulnerabilities...     os_version="15.4" pkg_num=370
2024-11-22T22:10:16+01:00       INFO    Number of language-specific files       num=1
2024-11-22T22:10:16+01:00       INFO    [gobinary] Detecting vulnerabilities...
2024-11-22T22:10:16+01:00       WARN    Using severities from other vendors for some vulnerabilities. Read https://aquasecurity.github.io/trivy/dev/docs/scanner/vulnerability#severity-selection for details.
2024-11-22T22:10:16+01:00       WARN    This OS version is no longer supported by the distribution      family="sles" version="15.4"
2024-11-22T22:10:16+01:00       WARN    The vulnerability detection may be insufficient because security updates are not provided
Vuln count: 487
```

File: Build/IntrospectGolang.pm [hunk 3]
```diff
@@ -168,6 +168,8 @@ sub buildinfo  {
   my ($fh) = @_;
   my ($vers, $mod) = rawbuildinfo($fh);
   return undef unless defined $vers;
+  $vers =~ s/^go(.*)/$1/;
+
```
Comment (mlschroe) [inline Build/IntrospectGolang.pm:172]: So I looked at the current syft code and they do not use the "go" prefix in the version of the synthetic "stdlib" package. So why exactly do we need to change this?

File: generate_sbom [hunk 4]
```diff
@@ -935,6 +942,18 @@ sub spdx_encode_header {
...
+  my ($dist) = @_;
+
+  return spdx_encode_pkg({
+    NAME => $dist->{id},
+    VERSION => $dist->{version_id},
+    spdx_id => sprintf('SPDXRef-OperatingSystem-%s', gen_pkg_id($dist)),
+    primaryPackagePurpose => 'OPERATING-SYSTEM',
+    skip_external_refs => 1
+  }, undef, undef, {});
+
```
Comment (mlschroe) [inline generate_sbom:955]: I think it's more clean to not call spdx_encode_pkg but just do the stuff we need for the operating system entry here.

File: Build/IntrospectGolang.pm (hunk 3 above)
Comment (josegomezr) [reply Build/IntrospectGolang.pm:172]: The prefix comes from the `rawbuildinfo` subroutine.

I've added a print just before the regex to inspect the outcome of the sub like:

```
# ...
  return undef unless defined $vers;
  use Data::Dumper;
  print(Dumper($vers, $mod));
  $vers =~ s/^go(.*)/$1/;
```

Then running it scanning a SLE-15SP4-Minimal-VM:
```
perl -I. generate_sbom --format cyclonedx --dir /mnt/ 
$VAR1 = 'go1.17.9';
$VAR2 = 'path	github.com/SUSE/connect-ng/suseconnect
mod	github.com/SUSE/connect-ng	(devel)	
';
```

`$vers` includes the go prefix that's not compliant with the version regex in trivy.

File: generate_sbom (hunk 4 above)
Comment (josegomezr) [reply generate_sbom:955]: Even if the filesystem entry is reflected _exactly_ as a package inside the OS?

**Review by mlschroe — APPROVED** [review]

**Pull request merged**